[WIP] feat: multiple media layers, moving

zhixiangteoh opened the pull request:
Changes:
1. Added grey intersecting point dot on toolbars as additional visual cue 
2. Added `MediaLayer`, `GlassLayer` classes
     - Each `MediaLayer` is tagged to one `Toolbar`
3. Added `createMediaLayer()` method in `MediaLayerManager` that creates `MediaLayer` objects
4. Updated `multiple-layers.js` to create separate independent video DOM elements per `MediaLayer`
     - Each layer's toolbar controls only affect that layer's video
5. [WIP] Create glass layers to handle moving and resizing

und3fined-v01d commented:
> 
> 
> Changes:
> 
>     1. Added grey intersecting point dot on toolbars as additional visual cue
> 
>     2. Added `MediaLayer`, `GlassLayer` classes
>        
>        * Each `MediaLayer` is tagged to one `Toolbar`
> 
>     3. Added `createMediaLayer()` method in `MediaLayerManager` that creates `MediaLayer` objects
> 
>     4. Updated `multiple-layers.js` to create separate independent video DOM elements per `MediaLayer`
>        
>        * Each layer's toolbar controls only affect that layer's video
> 
>     5. [WIP] Create glass layers to handle moving and resizing

This patch is doing too many things. Let's create a separate patch to handle moving

und3fined-v01d commented on util/Toolbar.js:
```diff
@@ -48,23 +46,21 @@ class Toolbar {
     }
 
-    createToolbarGroup(isAngled, { x, y, z }) {
+    createToolbarGroup(rotateXAngle, { position: { x, y, z } }) {
```
Destructure this within the function. It will help for error handling later

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -30,15 +29,14 @@ class App {
         this.video = this.createVideo(videoIn);
 
+        // Create Intersecting Point
+        this.intersectPoint = this.createIntersectPoint();
```
Let us remove this for now. There is already a highlight feature on toolbar intersection. We can add this later when needed.

zhixiangteoh replied:
I think this is helpful for quad layers when we want to arbitrarily place quad layers far down z axis and also move them arbitrarily. The controller ray should not be so long as to be able to reach them (e.g. the Oculus shell ray is short), and highlighting is not easy to see where you are pointing if the ray does not directly touch the toolbar.

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -54,12 +52,17 @@ class App {
-            this.toolbar.updateUI();
-        }
+        this.mediaLayers.forEach((mediaLayer) => {
+            mediaLayer.updateOnRender(xr.isPresenting);
```
Move the xr.isPresenting check outside the foor loop. This way we run the loop only if xr is presenting

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -54,12 +52,17 @@ class App {
-            this.toolbar.updateProgressBar();
+        for (const controller of this.controllers) {
+            if (this.mediaLayers.get("equirect")) {
+                this.handleToolbarIntersections(
```
This will not work if we have multiple equirect layers. We should modify this method to a more general case

zhixiangteoh replied:
This code for `equirect-buttons.js` is not the most updated; support for multiple layers (mainly quad layers) is in `multiple-layers.js`

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -69,16 +72,39 @@ class App {
+
             session.updateRenderState({
-                layers: [equirectLayer, session.renderState.layers[0]],
+                layers: [equirect.videoLayer, session.renderState.layers[0]],
```
Call it just layer instead of "videoLayer"

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -103,7 +129,7 @@ class App {
             sound.play();
 
-            this.handleToolbarIntersection(controller);
+            this.handleTriggerPress(controller);
```
Let's called this selectPress since thats the name of the default event too

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -138,7 +164,7 @@ class App {
         const line = new THREE.Line(geometry);
         line.name = "line";
-        line.scale.z = 10;
+        line.scale.z = 1;
```
Why has the scale been reduced?

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -217,39 +238,93 @@ class App {
     }
 
+    createIntersectPoint() {
+        const geometry = new THREE.CircleGeometry(0.02, 10);
```
Let's remove this for now

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -217,39 +238,93 @@ class App {
-            this.raycaster.ray.direction
-                .set(0, 0, -1)
-                .applyMatrix4(worldMatrix);
+    handleTriggerPress(controller) {
```
Name this handleSelectPress

und3fined-v01d commented on apps/equirect-buttons.js:
```diff
@@ -217,39 +238,93 @@ class App {
+
+        const intersections = this.raycaster.intersectObjects(objects);
+
+        let isAllToolbarsHidden = true;
```
this should be areAllToolbarsHidden

zhixiangteoh replied:
Makes sense.

zhixiangteoh commented on util/Toolbar.js:
```diff
@@ -1,25 +1,23 @@
-            y: 1.6,
-            z: -2,
-        });
+        this.toolbarGroup = this.createToolbarGroup(
```
Makes sense. In fact ui position does not really change, it's the toolbarGroup position that changes depending on where you want to place the whole toolbar.

zhixiangteoh commented:
> > Changes:
> > ```
> > 1. Added grey intersecting point dot on toolbars as additional visual cue
> > 
> > 2. Added `MediaLayer`, `GlassLayer` classes
> >    
> >    * Each `MediaLayer` is tagged to one `Toolbar`
> > 
> > 3. Added `createMediaLayer()` method in `MediaLayerManager` that creates `MediaLayer` objects
> > 
> > 4. Updated `multiple-layers.js` to create separate independent video DOM elements per `MediaLayer`
> >    
> >    * Each layer's toolbar controls only affect that layer's video
> > 
> > 5. [WIP] Create glass layers to handle moving and resizing
> > ```
> 
> This patch is doing too many things. Let's create a separate patch to handle moving

Agreed. Will split into different branches/PRs.

zhixiangteoh commented:
Closed by #24
